container.te

```diff
@@ -1483,6 +1483,8 @@ application_executable_file(kubelet_exec_t)
 can_exec(container_runtime_t, kubelet_exec_t)
 allow kubelet_t kubelet_exec_t:file entrypoint;
 
+container_lib_filetrans(kubelet_t,container_file_t,sock_file)
```

Comment:
It would be better to specify the name of the socket i.e. kubelet.sock as the 4th argument, to narrow scoped the rule affect, but when I tested it, it didn't work, neither  "kubelet.sock" nor "/var/lib/kubelet/pod-resources/kubelet.sock"

Reply:
Yes specify the forth parameter to make it tighter, but I am not sure we want this by default, since this means that any container that could get to the kublet.sock would be allowed to write to it.

Do we want this everywhere or just on a certain system where the container is supposed to talk to a kublet.sock, this is obviously not the default behavior we want for every kubernetes cluster in the world.

Reply:
Hey @rhatdan thank you for your reply, I would appreciate any hint about why specifying the name didn't work.
When I tested it locally I was using the underlying command the `container_lib_filetrans`  macro is using:
```
(typetransition kubelet_t container_var_lib_t sock_file "/var/lib/kubelet/pod-resources/kubelet.sock" container_file_t)
```
 and it didn't work

> since this means that any container that could get to the kublet.sock would be allowed to write to it.

The pod-resource socket has a very specific use and it exposes API only for reading information from Kubelet and writes nothing (the write permission are for writing the initial request for getting that information). 

> Do we want this everywhere or just on a certain system where the container is supposed to talk to a kublet.sock, this is obviously not the default behavior we want for every kubernetes cluster in the world.

We mainly care about the OCP case, since on other K8S flavors users can manage the SELinux permissions as they see fit.

Reply:
I tried:
```
(typetransition kubelet_t container_var_lib_t sock_file kubelet.sock container_file_t
```
as well and it didn't work
Am I missing something here?

Reply:
Nope that looks right but you need to make sure all of them are correct.  Remove the kublet_sock from the system, make sure that kublet is running as kublet_t when it creats the socket, make sure the parent directory is labeled container_varlib_t.

If all of those things are correct it will work.

Reply:
Ok that was a typo by me.

You could write a script which creates a kubelet.sock socket in a directory labeled container_file_t and see it and run it via the runcom command above to see if this works correctly.

Reply:
This is all the data I extracted from the node:

```
[root@cnfdf08 targeted]# ls -Z /usr/bin/kubelet 
system_u:object_r:kubelet_exec_t:s0 /usr/bin/kubelet

[root@cnfdf08 targeted]# ps -eZ | grep kubelet
system_u:system_r:kubelet_t:s0-s0:c0.c1023 3920522 ? 1-02:16:38 kubelet

# parent dir
[root@cnfdf08 targeted]# ls -lZ /var/lib/kubelet/ | grep pod-r
drwxr-x---.  2 root root system_u:object_r:container_var_lib_t:s0   26 Sep  3 14:16 pod-resources

```
And unfortunately it doesn't work

Reply:
```
$ sesearch  -T -s kubelet_t -t container_var_lib_t
type_transition kubelet_t container_var_lib_t:dir container_file_t _data;
type_transition kubelet_t container_var_lib_t:dir container_file_t snapshots;
type_transition kubelet_t container_var_lib_t:dir container_ro_file_t atomic;
type_transition kubelet_t container_var_lib_t:dir container_ro_file_t containerd-btrfs;
type_transition kubelet_t container_var_lib_t:dir container_ro_file_t containerd-fuse-overlayfs;
type_transition kubelet_t container_var_lib_t:dir container_ro_file_t containerd-native;
type_transition kubelet_t container_var_lib_t:dir container_ro_file_t containerd-nydus;
type_transition kubelet_t container_var_lib_t:dir container_ro_file_t containerd-overlaybd;
type_transition kubelet_t container_var_lib_t:dir container_ro_file_t containerd-overlayfs;
type_transition kubelet_t container_var_lib_t:dir container_ro_file_t containerd-soci;
type_transition kubelet_t container_var_lib_t:dir container_ro_file_t containerd-stargz;
type_transition kubelet_t container_var_lib_t:dir container_ro_file_t containerd-zfs;
type_transition kubelet_t container_var_lib_t:dir container_ro_file_t executor;
type_transition kubelet_t container_var_lib_t:dir container_ro_file_t init;
type_transition kubelet_t container_var_lib_t:dir container_ro_file_t kata-containers;
type_transition kubelet_t container_var_lib_t:dir container_ro_file_t overlay-images;
type_transition kubelet_t container_var_lib_t:dir container_ro_file_t overlay-layers;
type_transition kubelet_t container_var_lib_t:dir container_ro_file_t overlay2-images;
type_transition kubelet_t container_var_lib_t:dir container_ro_file_t overlay2-layers;
type_transition kubelet_t container_var_lib_t:dir container_ro_file_t overlay2;
type_transition kubelet_t container_var_lib_t:dir container_ro_file_t overlay;
type_transition kubelet_t container_var_lib_t:dir container_ro_file_t sandboxes;
type_transition kubelet_t container_var_lib_t:file container_log_t container-json.log;
type_transition kubelet_t container_var_lib_t:file container_ro_file_t config.env;
type_transition kubelet_t container_var_lib_t:file container_ro_file_t hostname;
type_transition kubelet_t container_var_lib_t:file container_ro_file_t hosts;
```

Reply:
is there some dryrun command or logs that can indicate if there's a problem?

Reply:
Did you install the policy?  If you do could you check if it shows in the seinfo line above?  I have not tried this locally.

Reply:
After installing it I see:

$ sesearch  -T -s kubelet_t -t container_var_lib_t -c sock_file
type_transition kubelet_t container_var_lib_t:sock_file container_file_t;

which looks good.

Reply:
No I just changed the current policy, added the line and reinstall with:
`sudo semodule -i container.cil`

Reply:
```
[root@cnfdf08 ~]# sesearch  -T -s kubelet_t -t container_var_lib_t
-bash: sesearch: command not found
```

Reply:
I'm checking inside OCP node, I cannot install packages using dnf

Reply:
I just checked it locally with this script

```
$ cat > /tmp/kubelet.sh
#/bin/sh
id -Z
tmpdir=$(mktemp -d)
chcon -t container_var_lib_t ${tmpdir}
timeout 1s ncat -l -U ${tmpdir}/kubelet.sock
ls -lZ ${tmpdir}/kubelet.sock
rm -rf ${tmpdir}
_EOF
$ chmod +x /tmp/kubelet.sh
$ sudo setenforce 0
$ runcon system_u:system_r:kubelet_t:s0 /tmp/kubelet.sh 
system_u:system_r:kubelet_t:s0
srwxr-xr-x. 1 dwalsh dwalsh system_u:object_r:container_file_t:s0 0 Sep 12 12:06 /tmp/tmp.riSH1GZFOF/kubelet.sock
$ sudo setenforce 1
```

Reply:
If that script works on your system, then the issue is in how you are testing it.

Reply:
> After installing it I see:

$ sesearch -T -s kubelet_t -t container_var_lib_t -c sock_file
type_transition kubelet_t container_var_lib_t:sock_file container_file_t;

which looks good.

Yes but this is without the explicit name.

can you try:

container_lib_filetrans(kubelet_t,container_file_t,sock_file,kubelet.sock)

Adding the name as the forth argument

Reply:
Made the change:

```
$ git diff
diff --git a/container.te b/container.te
index 989a026..6a321be 100644
--- a/container.te
+++ b/container.te
@@ -1508,6 +1508,7 @@ optional_policy(`
        unconfined_run_to(kubelet_t, kubelet_exec_t)
        role_transition unconfined_r kubelet_exec_t system_r;
 ')
+container_lib_filetrans(kubelet_t,container_file_t,sock_file, "kubelet.sock")
 
 # Standard container which needs to be allowed to use any device
 container_domain_template(container_device, container)

$ sesearch  -T -s kubelet_t -t container_var_lib_t -c sock_file
type_transition kubelet_t container_var_lib_t:sock_file container_file_t kubelet.sock;
$ runcon system_u:system_r:kubelet_t:s0 /tmp/kubelet.sh 
system_u:system_r:kubelet_t:s0
srwxr-xr-x. 1 dwalsh dwalsh system_u:object_r:container_file_t:s0 0 Sep 12 12:13 /tmp/tmp.diPYppwAzx/kubelet.sock
```

Reply:
> You could write a script which creates a kubelet.sock socket in a directory labeled container_file_t and see it and run it via the runcom command above to see if this works correctly.

The parent directory should be labels as  container_file_t as well?

Reply:
Modified the script to also create notkubelt.sock.

```
cat /tmp/kubelet.sh 
#/bin/sh
id -Z
tmpdir=$(mktemp -d)
chcon -t container_var_lib_t ${tmpdir}
timeout 1s ncat -l -U ${tmpdir}/kubelet.sock
timeout 1s ncat -l -U ${tmpdir}/notkubelet.sock
ls -lZ ${tmpdir}
rm -rf ${tmpdir}

runcon system_u:system_r:kubelet_t:s0 /tmp/kubelet.sh 
system_u:system_r:kubelet_t:s0
total 0
srwxr-xr-x. 1 dwalsh dwalsh system_u:object_r:container_file_t:s0    0 Sep 12 12:15 kubelet.sock
srwxr-xr-x. 1 dwalsh dwalsh system_u:object_r:container_var_lib_t:s0 0 Sep 12 12:15 notkubelet.sock
```

Reply:
cool so it suppose to work. great. i'll modify the PR then
\

Reply:
The parent directory should be labeled container_var_lib_t. Not container_file_t.

Reply:
> The parent directory should be labeled container_var_lib_t. Not container_file_t.

Then it's ok this is the current state.

I just updated the PR please take a final look and see if you're happy with it.

Reply:
Where does kubelet usually create its kubelet.sock?  Is it in a container_var_lib_t directory, then I would not like this patch because it is too loose and still will not allow what you want.

$  sesearch -A -s container_t -t kubelet_t -p connectto

This means that if you leak the socket file into a container the container can not write to it and can not communicate with the kubelet.

From a security point of view allow all containers to communicate with the kublet is a disaster, if it can get the kublet to act on its behalf.

If you want only a single container to communicate with the kubelet then we should write a new policy for the containier (container_kubelet_t) for example.  Then only this container type would be allowed to communicate with the kubelet, we would then also change the type of the sock_file to something other containers can not write to by default.

Reply:
The plan is to set my container as a container_device_plugin_t and not container_t which has an access to kubelet socket.

This way standard container_t won't be able to access the socket even if they'll mount it.

My container also need an access to the api server and this is the reason why we were needed #325

Reply:
Also, there was already a try to achieve the same as we are doing here in https://github.com/containers/container-selinux/pull/275 
But it didn't work because kubelet creates the socket at runtime. This is the reason why the current PR is needed.

Reply:
Then why changes the kubelet.sock at all, just allow container_device_plugin_t to access the label of the kubelet.sock

Reply:
Allow container_device_plugin_t to access container_var_lib_t?

Would you prefer this approach?

Reply:
No I would rather that kublet.sock be labeled kublet_var_run_t. or kublet_var_lib_t.

Reply:
kubernetes_file_t

Reply:
OK I updated the PR.
FYI I see that for the other sockets like the device plugin socket, the label is `container_file_t`:
```
[root@cnfdf08 ~]# ls -Z /var/lib/kubelet/device-plugins/kubelet.sock 
system_u:object_r:container_file_t:s0 /var/lib/kubelet/device-plugins/kubelet.sock
```

Which is done by Kubelet itself:
https://github.com/kubernetes/kubernetes/blob/5ac315faf428567ec023aa7d8c2b89b1d9c9211a/pkg/kubelet/kubelet.go#L1428 
I'm adding this information in order to provide a complete picture

Reply:
Thinking about it, this might also change the device plugins socket's label I mentioned above to kubernetes_file_t, because their permissions and names are exactly the same, isn't it?

Reply:
Yup either way.  The kubelet change above is a potential security risk.  Allowing all container processes to talk to your socket, means that you are opening yourself up to attack from everyone.  Changing to only allow a specific type to communicate is much more secure.

Reply:
Lets setup a specific audio call, so we can finish this up.